### ISSUE TITLE
logs would be more useful if they were structured (go-hclog migration)

### ISSUE THREAD

**Issue by macb** [hashicorp/consul#1249]
The logging in consul is quite useful already, especially when exploring bits and pieces with debug/trace. However, the format is really only useful to humans and/or programs intimately familiar with consul's logs. It would be incredibly useful if the format was configurable and one such format offered more computer parseable structure.

As an example:

https://github.com/hashicorp/consul/blob/master/command/agent/watch_handler.go#L43

```
logger.Printf("[ERR] agent: Failed to setup watch: %v", err)
```

This line contains a lot of information:
- The level: We're logging an error.
- Where we are: The error is coming from the agent subsystem
- The impact: We failed to setup a watch
- What happened: The error itself.

In order to parse this log line we'd need to strip off the level, pull out what system we're in, take out the identifiable string ("failed to setup watch") and also store the error that arose. This is relatively easy to do as humans but computers are a little less flexible.

If instead we used a structured logging format like JSON we could end up with a log like:

``` json
{
  "level": "error",
  "system": "agent",
  "error": "the error string from the error that occured",
  "msg": "failed to setup watch" 
}
```

And to produce that output we could change our interface accordingly:

```
// setup early and re-used as needed
logger := log.WithField("system", "agent")
// Do some stuff
if err != nil {
    logger.WithError(err).Error("failed to setup watch")
}
```

Breaking this interface down: 
- `.WithField` could add an arbitrary k/v pair to the log. It returns a new structure so that we could chain them together without altering the parent.
- `.WithError` might be a convenience method for consistencies sake and creates an `"error"` key with the value being the error's value.
- `.Error` is the print method itself which serves two purposes. It creates the "message" we want to send (the consistent string we can depend on always existing in this log line) as well as setup the level that we want to print at. The underlying writer interface could then be responsible for deciding whether or not we should _actually_ print the log based on the log level we're configured to.

On top of homegrown solutions structured logs can play quite nicely with systems like [rsyslog](http://www.rsyslog.com/tag/cee/). A very popular generic library for this flow is https://github.com/Sirupsen/logrus. There's no doubt this would be a pretty big change which is why I wanted to open the issue and get input before going rogue.

**Comment by slackpad** [comment]
Hi @macb - thanks for opening this. I think we are going to take a look at doing something along these lines across more than just Consul, so you should probably hold off on "going rogue" :-) We will update this once we've got a plan.

**Comment by macb** [comment]
Awesome @slackpad, looking forward to it.

**Comment by bastichelaar** [comment]
Any progress on this? Structured logging would indeed be very useful!

**Comment by slackpad** [comment]
Hi @bastichelaar still nothing yet, we've had too many other things ahead of this on the roadmap. I'll ping this up internally and see if we can get some discussion going.

**Comment by bastichelaar** [comment]
Thanks!

**Comment by blalor** [comment]
@slackpad if you can come up with a recommendation (I'm personally a fan of [logrus](github.com/Sirupsen/logrus)), that might be the kind of thing a community contributor could tackle.  I'm personally irked by the logging output of consul-template, consul, and vault; if you could settle on a framework across all your products, that'd be a really great experience for us ops-y folk!

**Comment by sean-** [comment]
Thanks for that recommendation @blalor.  It's something that has come up several times internally and isn't a resolved issue yet.  Stay tuned.

**Comment by GMartinez-Sisti** [comment]
Hi!

I'm also interested in having an option to specify logs in JSON. This will be extremely useful in logging systems that can parse JSON, such as Filebeat/Logstash.

Do you have any news on this?

Regards

**Comment by jippi** [comment]
Nomad started to use [go-hclog](https://github.com/hashicorp/go-hclog) - so i would expect Consul / Vault and other tools to soon'ish adopt this as well

**Comment by pearkes** [comment]
As @jippi correctly pointed out we're going to adopt https://github.com/hashicorp/go-hclog. We plan to do this in an upcoming major release.

**Comment by s-christoff** [comment]
Howdy! We have implemented hclog with structured logging in #7130 - I am going to close this issue for now. :)
